channel/erc20_depositor.go

```diff
@@ -40,42 +45,138 @@ const ERC20DepositorTXGasLimit = 100000
+//
+//nolint:funlen
 func (d *ERC20Depositor) Deposit(ctx context.Context, req DepositReq) (types.Transactions, error) {
-	// Bind a `AssetHolderERC20` instance.
-	assetholder, err := assetholdererc20.NewAssetholdererc20(req.Asset.EthAddress(), req.CB)
-	if err != nil {
-		return nil, errors.Wrapf(err, "binding AssetHolderERC20 contract at: %x", req.Asset)
-	}
+	lockKey := lockKey(req.Account.Address, req.Asset.EthAddress())
+	lock := handleLock(lockKey)
+
 	// Bind an `ERC20` instance.
 	token, err := peruntoken.NewPeruntoken(d.Token, req.CB)
 	if err != nil {
 		return nil, errors.Wrapf(err, "binding ERC20 contract at: %x", d.Token)
 	}
-	// Increase the allowance.
-	opts, err := req.CB.NewTransactor(ctx, ERC20DepositorTXGasLimit, req.Account)
-	if err != nil {
-		return nil, errors.WithMessagef(err, "creating transactor for asset: %x", req.Asset)
+	callOpts := bind.CallOpts{
+		Pending: false,
+		Context: ctx,
 	}
-	tx1, err := token.IncreaseAllowance(opts, req.Asset.EthAddress(), req.Balance)
+	// variables for the return value.
+	var depResult DepositResult
+	var approvalReceived bool
+	var tx1 *types.Transaction
+	var err1 error
+	lockAndUnlock(lock, func() {
+		allowance, err := token.Allowance(&callOpts, req.Account.Address, req.Asset.EthAddress())
+		if err != nil {
+			depResult.Transactions = nil
+			depResult.Error = errors.WithMessagef(err, "could not get Allowance for asset: %x", req.Asset)
+		}
+		result := new(big.Int).Add(req.Balance, allowance)
+
+		// Increase the allowance.
+		opts, err := req.CB.NewTransactor(ctx, ERC20DepositorTXGasLimit, req.Account)
+		if err != nil {
+			depResult.Transactions = nil
+			depResult.Error = errors.WithMessagef(err, "creating transactor for asset: %x", req.Asset)
+		}
+		// Create a channel for receiving PeruntokenApproval events
+		eventSink := make(chan *peruntoken.PeruntokenApproval)
+
+		// Create a channel for receiving the Approval event
+		eventReceived := make(chan bool)
+
+		// Watch for Approval events and send them to the eventSink
+		subscription, err := token.WatchApproval(&bind.WatchOpts{Start: nil, Context: ctx}, eventSink, []common.Address{req.Account.Address}, []common.Address{req.Asset.EthAddress()})
+		if err != nil {
+			depResult.Transactions = nil
+			depResult.Error = errors.WithMessagef(err, "Cannot listen for event")
+		}
+		tx1, err1 = token.Approve(opts, req.Asset.EthAddress(), result)
+		if err1 != nil {
+			err = cherrors.CheckIsChainNotReachableError(err)
+			depResult.Transactions = nil
+			depResult.Error = errors.WithMessagef(err, "increasing allowance for asset: %x", req.Asset)
+		}
+
+		go func() {
+			select {
+			case event := <-eventSink:
+				log.Printf("Received Approval event: Owner: %s, Spender: %s, Value: %s\n", event.Owner.Hex(), event.Spender.Hex(), event.Value.String())
+				eventReceived <- true
+			case err := <-subscription.Err():
+				log.Println("Subscription error:", err)
+			}
+		}()
+		approvalReceived = <-eventReceived
+	})
```

Comment:
I would definitely put this code into a separate method which then also removes the need of the `lockAndUnlock` function and for the `nolint:funlen` exception.

Reply:
I think they used to call what lock and unlock does a "critical section". Although it may also have had other implications regarding elevated process priority, not sure.